keps/sig-storage/1847-autoremove-statefulset-pvcs/README.md

```diff
@@ -0,0 +1,590 @@
+and RemoveOnScaleDown and the version of the kube-controller downgraded,
+even though the `PersistentVolumeClaimReclaimPolicy` field will go away, the references
+would still be acted upon by the garbage collector and cleaned up 
+based on the settings before downgrade. 
+
+<!--
+If applicable, how will the component be upgraded and downgraded? Make sure
+this is in the test plan.
+
+Consider the following in developing an upgrade/downgrade strategy for this
+enhancement:
+- What changes (in invocations, configurations, API use, etc.) is an existing
+  cluster required to make on upgrade in order to keep previous behavior?
+- What changes (in invocations, configurations, API use, etc.) is an existing
+  cluster required to make on upgrade in order to make use of the enhancement?
+-->
+
+### Version Skew Strategy
+There is only kubecontroller manager changes involved, hence not applicable for
+version skew involving other components.
+
+<!--
+If applicable, how will the component handle version skew with other
+components? What are the guarantees? Make sure this is in the test plan.
+
+Consider the following in developing a version skew strategy for this
+enhancement:
+- Does this enhancement involve coordinating behavior in the control plane and
+  in the kubelet? How does an n-2 kubelet without this feature available behave
+  when this feature is used?
+- Will any other components on the node change? For example, changes to CSI,
+  CRI or CNI may require updating that component before the kubelet.
+-->
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+* **How can this feature be enabled / disabled in a live cluster?**
+  - [x] Feature gate (also fill in values in `kep.yaml`)
+    - Feature gate name: statefulset-autodelete-pvcs
+    - Components depending on the feature gate: kube-controller-manager
+  
+* **Does enabling the feature change any default behavior?**
+  The default behaviour is only changed when user explicitly specifies the `PersistentVolumeClaimReclaimPolicy`. 
+  Hence no change in any user visible behaviour change by default.
+
+* **Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?**
+  Yes, but with side effects for users who already started using the feature by means of 
+  specifying non-retain `PersistentVolumeClaimReclaimPolicy`. We will an annotation to the
+  PVC indicating that the references have been set from previous enablement. Hence a reconcile
+  loop which goes through the requried PVCs and removes the references will be added. 
+  The side effect is that if there was pod deletion before the references were removed after the
+  feature flag was diabled, the PVCs could get deleted.
+  
+* **What happens if we reenable the feature if it was previously rolled back?** 
+The reconcile loop which removes references on disablement will not come into action. Since the 
```

Comment:
This should have no effect, as the policy fields were removed during the rollback, so when reenabled all StatefulSets in the cluster will be using the retain policy.

Reply:
The way feature disabling/rollback works is that the field [remains stored in etcd](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#alpha-field-in-existing-api-version), however controllers will not process the field (because it's protected by a feature gate). Then if you re-enable the feature again, then you can start processing the field again.

Reply:
Thanks.

I think that means that the RemoveOnDeletion needs to be reconciled, that is make sure associated PVCs have ownerrefs to the statefulset.

I don't see this as too much of a problem if things aren't exactly consistent. For example, one might set RemoveOnScaledown on a 10 replica statefulset, disable the feature, scale down to 6 pods, then re-enable the feature; PVCs 7-10 are going to not have been deleted and we can't be expected to delete them. So if the disabling or re-enabling happens during a scaledown and some PVCs are missed or deleted, the behavior is almost the same as before when we were okay with the orphaned PVCs 7-10. 

The main thing I think is to make sure "future" scaledown/deletions work correctly, and as long as we get the statefulset ownerref for RemoveOnDeletion that will be covered.